feat: TOOLS-3077 - Add Server Diff Functionality to asconfig diff Command

a-spiker opened the pull request:
### **Summary**

This PR significantly enhances the `asconfig diff` command by:
- Adding support for comparing a local config file against a live Aerospike server (`diff server` subcommand).

---

### **Key Changes**

#### **1. Feature: `diff server` Subcommand**
- Adds `asconfig diff server [flags] <path/to/config>` to compare a local config file (YAML or .conf) with the live configuration of a running Aerospike server.
- Adds `asconfig diff files [flags] <path/to/config1> <path/to/config2>` to compare two config file (YAML or .conf) 
- asconfig diff <path/to/config1> <path/to/config2> still is supported but throws a warning to use `asconfig diff files`
- Supports all necessary Aerospike connection flags.
- Provides clear help text and usage examples.
- Marks the feature as BETA and hides the `--format` flag from help output from diff server (auto-detected or settable for advanced use).

#### **2. CLI Help Improvements**
- Updates the top-level `diff` command help to mention both file-to-file and file-to-server comparison as subcommands.
- Clarifies argument requirements and format detection in both command and subcommand help texts.

#### **4. Test Coverage**
- Adds and refactors unit tests for:
  - Argument validation (too few/many args, non-existent files, etc.)
  - Format detection (flag vs. extension, invalid formats)
  - Edge cases for both file and server diff modes
  - Flat map diff logic (including logging enum and metadata handling)
- Ensures robust error handling and clear test descriptions.

#### **5. Refactoring and Logging**
- Refactors core diff logic into `runFileDiff` and `runServerDiff` for clarity and maintainability.
- Adds debug logging for key steps and error cases.

---

### **Motivation**

- Users need to compare local config files with the live state of Aerospike servers for troubleshooting and validation.
- Consistent and predictable format handling is critical for usability.

---

### **How to Use**

- **Compare two files: (Deprecated) for backward compatibility **
  ```
  asconfig diff file1.yaml file2.yaml
  ```
- **Compare two files:**
  ```
  asconfig diff files file1.yaml file2.yaml
  ```  
- **Compare a file with a running server:**
  ```
  asconfig diff server -h <host> -U <user> -P <password> file.conf
  ```
  (or with YAML: `asconfig diff server  ... file.yaml`)  

---

### **Notes**

- The `diff server` feature is marked as BETA as it uses generate functionality, which is in BETA
- All changes are backward compatible with previous `diff` usage.

---

a-spiker committed:
feat: add server diff functionality to compare local config with running Aerospike server

- Introduced a new `--server` flag for the `diff` command to allow comparison of a local configuration file against the live configuration from a running Aerospike server.
- Updated error handling for argument counts specific to server diff.
- Refactored the diff command to separate file-to-file and server-to-file comparison logic.

a-spiker committed:
test: add comprehensive tests for server diff functionality and argument validation

- Introduced new test cases for the `--server` flag in the diff command, covering various scenarios including missing arguments, invalid formats, and non-existent files.
- Added validation tests for server diff arguments to ensure proper error handling.
- Enhanced existing tests for file diff functionality to include edge cases and argument validation.

a-spiker committed:
chore: update go.sum to remove obsolete aerospike-management-lib entries

- Removed outdated versions of aerospike-management-lib from go.sum to streamline dependencies and ensure compatibility with the latest library versions.

a-spiker committed:
Merge branch 'main' into TOOLS-3077

a-spiker committed:
feat: Improve type-aware comparison in diff logic

Replaces reflect.DeepEqual with a custom valuesEqual function for more robust type-aware comparison, handling cases like int vs string, slices, and booleans. Ensures server and local configs are parsed through the same path to normalize data types before diffing, improving accuracy of configuration diffs.

a-spiker committed:
fix: Improve valuesEqual logic and add comprehensive tests

Refactored the valuesEqual function to compare values by string conversion first, improving type flexibility and preventing panics on slice comparison. Added extensive unit tests for valuesEqual, isSlice, slicesEqual, diffFlatMaps, and argument validation in diff_test.go to ensure correctness and robustness of diff logic.

a-spiker committed:
feat: make slice comparison order-agnostic and efficient

Refactored slicesEqual to compare slices in an order-agnostic way using frequency counting for O(n) complexity. Added convertToStringSlice to handle different slice types and ensure consistent comparison. Updated tests to validate new behavior, including mixed types and order variations.

a-spiker commented on cmd/diff.go:
```diff
@@ -167,10 +134,222 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
+
+// valuesEqual compares two values with type-aware comparison
+// This handles cases where values might be the same but different types
+// (e.g., int vs string, int vs int64, slices, etc.)
```
Had assumed that slices would be in the same order during parsing. Made it order agnostic.

a-spiker replied:
https://go.dev/play/p/zmnP8QyevYl

During our parsing, if it fails.. thought of defensive programming.

a-spiker commented on cmd/diff.go:
```diff
@@ -167,10 +134,222 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
+}
+
+// slicesEqual compares two values that might be slices
+func slicesEqual(v1, v2 any) bool {
```
Had assumed that slices would be in the same order during parsing. Made it order agnostic.

a-spiker committed:
feat: Add beta warning and adjust log levels in diff command

Introduces a warning message indicating the server diff feature is in beta

Since the generate command is in beta and diff with server uses generate functionality

dwelch-spike reviewed (changes requested):
Overall this looks like a nice way to make comparing server configs easier 👍 I think we need to think about this feature some more before adding it though. Here are the 2 main things on my mind

1. I'm not sure about the type agnostic value equality checking code added. I think it might hide errors in the parsing code and potentially miss valid diffs, I'd hold off on adding code like that until a real issue is observed in the field or testing. I'm open to being convinced otherwise though

2. A nitpick, I don't like that the help output for `diff` includes all the cluster connection options by default now. When you are just diffing 2 files, none of them are needed.
See below
```
./asconfig diff --help
Diff is used to compare differences between Aerospike configuration files.
                                It is used on two files of the same format from any format
                                supported by the asconfig tool, e.g. yaml or Aerospike config.
                                Schema validation is not performed on either file. The file names must end with
                                extensions signifying their formats, e.g. .conf or .yaml, or --format must be used.

                                When using --server flag, compare a local configuration file against
                                the live configuration from a running Aerospike server. In this mode,
                                only one config file path is required as an argument.

                                Examples:
                                  # Compare two local configuration files
                                  asconfig diff aerospike1.yaml aerospike2.yaml
                                  
                                  # Compare local config against running server
                                  asconfig diff --server -h 127.0.0.1 -U admin -P admin aerospike.yaml
                                  
                                  # Compare local config against server with TLS
                                  asconfig diff -s -h aerospike.example.com --tls-enable aerospike.conf

Usage:
  asconfig diff [flags] <path/to/config1> <path/to/config2>

Flags:
      --auth INTERNAL,EXTERNAL,PKI                                                                  The authentication mode used by the Aerospike server. INTERNAL 
                                                                                                    uses standard user/pass. EXTERNAL uses external methods (like LDAP) 
                                                                                                    which are configured on the server. EXTERNAL requires TLS. PKI allows 
                                                                                                    TLS authentication and authorization based on a certificate. No 
                                                                                                    username needs to be configured. (default INTERNAL)
  -F, --format string                                                                               The format of the source file(s). Valid options are: yaml, yml, and conf. (default "conf")
  -h, --host host[:tls-name][:port][,...]                                                           The Aerospike host. (default 127.0.0.1)
  -P, --password "env-b64:<env-var>,b64:<b64-pass>,file:<pass-file>,<clear-pass>"                   The Aerospike password to use to connect to the Aerospike 
                                                                                                    cluster.
  -p, --port int                                                                                    The default Aerospike port. (default 3000)
  -s, --server                                                                                      Compare local config file against a running Aerospike server configuration.
      --tls-cafile env-b64:<cert>,b64:<cert>,<cert-file-name>                                       The CA used when connecting to Aerospike.
      --tls-capath <cert-path-name>                                                                 A path containing CAs for connecting to Aerospike.
      --tls-certfile env-b64:<cert>,b64:<cert>,<cert-file-name>                                     The certificate file for mutual TLS authentication with 
                                                                                                    Aerospike.
      --tls-enable                                                                                  Enable TLS authentication with Aerospike. If false, other TLS 
                                                                                                    options are ignored.
      --tls-keyfile env-b64:<cert>,b64:<cert>,<cert-file-name>                                      The key file used for mutual TLS authentication with Aerospike.
      --tls-keyfile-password "env-b64:<env-var>,b64:<b64-pass>,file:<pass-file>,<clear-pass>"       The password used to decrypt the key file if encrypted.
      --tls-name string                                                                             The server TLS context to use to authenticate the connection to 
                                                                                                    Aerospike.
      --tls-protocols "[[+][-]all] [[+][-]TLSv1] [[+][-]TLSv1.1] [[+][-]TLSv1.2] [[+][-]TLSv1.3]"   Set the TLS protocol selection criteria. This format is the same 
                                                                                                    as Apache's SSLProtocol documented at 
                                                                                                    https://httpd.apache.org/docs/current/mod/mod_ssl.html#ssl protocol. (default +TLSv1.2)
  -U, --user string                                                                                 The Aerospike user to use to connect to the Aerospike cluster.

Global Flags:
      --config-file string   Config file (default is /etc/aerospike/astools)
  -u, --help                 Display help information
      --instance string      For support of the aerospike tools toml schema. Sections with the 
                             instance are read. e.g in the case where instance 'a' is specified 
                             sections 'cluster_a', 'uda_a' are read.
  -l, --log-level string     Set the logging detail level. Valid levels are: [panic fatal error warning info debug trace] (default "info")
  -V, --version              Display version.
```
I think you can make it cleaner by adding the cluster flags to their own group with something like this in the newDiffCmd function
```
	// Group server-specific flags for cleaner help output
	res.SetHelpFunc(func(cmd *cobra.Command, args []string) {
		// Group flags for cleaner help output
		fg := cmd.Flags()
		serverFg := pflag.NewFlagSet("server", pflag.ContinueOnError)
		serverFg.AddFlagSet(asFlagSet)

		// Create dedicated flag groups
		requiredFlags := pflag.NewFlagSet("required", pflag.ContinueOnError)
		requiredFlags.AddFlag(fg.Lookup("format"))
		requiredFlags.AddFlag(fg.Lookup("server"))

		// Print default help, but with custom flag groupings
		fmt.Fprintf(cmd.OutOrStdout(), "%s\n\n", cmd.Long)
		fmt.Fprintf(cmd.OutOrStdout(), "Usage:\n  %s\n\n", cmd.Use)
		fmt.Fprintf(cmd.OutOrStdout(), "Flags:\n%s\n", requiredFlags.FlagUsages())
		fmt.Fprintf(cmd.OutOrStdout(), "Server Connection Flags (for use with --server):\n%s\n", serverFg.FlagUsages())
		fmt.Fprintf(cmd.OutOrStdout(), "Global Flags:\n%s\n", cmd.InheritedFlags().FlagUsages())
	})
```

dwelch-spike commented:
I played around with making the `--server` flag a subcommand that would be used like `asconfig diff server [flags] <local/path>` and the help looks pretty clean. I think this would be a clean way to organize it. @a-spiker what do you think?

```
 ./asconfig diff --help
Diff is used to compare differences between Aerospike configuration files.
                                It is used on two files of the same format from any format
                                supported by the asconfig tool, e.g. yaml or Aerospike config.
                                Schema validation is not performed on either file. The file names must end with
                                extensions signifying their formats, e.g. .conf or .yaml, or --format must be used.

                                To compare a local file against a running server, see the 'server' subcommand.

                                Examples:
                                  # Compare two local configuration files
                                  asconfig diff aerospike1.yaml aerospike2.yaml

Usage:
  asconfig diff <path/to/config1> <path/to/config2> [flags]
  asconfig diff [command]

Available Commands:
  server      Compare a local config file against a running Aerospike server.

Flags:
  -F, --format string   The format of the source file(s). Valid options are: yaml, yml, and conf. (default "conf")

Global Flags:
      --config-file string   Config file (default is /etc/aerospike/astools)
  -u, --help                 Display help information
      --instance string      For support of the aerospike tools toml schema. Sections with the 
                             instance are read. e.g in the case where instance 'a' is specified 
                             sections 'cluster_a', 'uda_a' are read.
  -l, --log-level string     Set the logging detail level. Valid levels are: [panic fatal error warning info debug trace] (default "info")
  -V, --version              Display version.

Use "asconfig diff [command] --help" for more information about a command.
```

a-spiker committed:
refactor: refactor diff command and add subcommand for server diff

Refactors the diff command to use a subcommand structure, moving server diff functionality to a new 'diff server' subcommand. Simplifies flag handling, updates help text and examples, and streamlines the diff logic by removing custom value comparison in favor of reflect.DeepEqual.

a-spiker committed:
refactor: Remove redundant and low-level unit tests from diff_test.go

This commit removes extensive low-level unit tests for helper functions such as valuesEqual, isSlice, and slicesEqual, as well as redundant server diff argument validation and flag validation tests. The remaining tests focus on higher-level command behavior and integration, streamlining the test suite and reducing maintenance overhead.

a-spiker committed:
docs: Update diff command descriptions for server comparison

Expanded the short and long descriptions of the diff command to clarify support for comparing a config file against a running server. Marked the server comparison feature as BETA in the server subcommand descriptions.

a-spiker committed:
docs: TOOLS-3077 Update diff command descriptions for clarity

Improved the short and long descriptions for the diff and server subcommands to clarify that comparisons are made against a running server's configuration. Updated examples and added notes about supported file formats for better user guidance.

a-spiker committed:
fix: Refactor format flag handling in diff commands

Moved the --format flag from persistent to local flags in the diff command and hid it in the server diff subcommand, as the format will be auto-detected. Added a debug log to show the detected local file format.

a-spiker commented on cmd/diff.go:
```diff
@@ -167,10 +135,287 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
+
+	// Convert both to strings and compare first
+	str1 := fmt.Sprintf("%v", v1)
+	str2 := fmt.Sprintf("%v", v2)
```
Removed the extra logic around comparing

a-spiker commented on cmd/diff.go:
```diff
@@ -167,10 +135,287 @@ func diffFlatMaps(m1 map[string]any, m2 map[string]any) []string {
+
+	// Use reflection to check if it's actually a slice
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Slice {
```
Removed the extra logic around comparing

a-spiker committed:
feat: TOOLS-3077 Refactor diff command to use subcommands

Introduces 'files' and 'server' subcommands for the diff command, improving clarity and extensibility. The legacy behavior is preserved for backward compatibility, with a warning message. Adds tests for both legacy and new subcommand usage, including argument validation.

a-spiker committed:
fix: Update diff tests to use substring match

Changed the diff test assertion to check if the output contains the expected result instead of requiring an exact match. This allows for additional warnings or messages in the output without failing the test.

a-spiker committed:
docs: Update diff command help text and examples

Improved the help text and examples for the diff, diff files, and diff server commands to clarify usage, supported formats, and default behaviors. Added more detailed flag descriptions and updated example host/port formats for consistency.

dwelch-spike reviewed (dismissed):
LGTM I'm happy with the `diff files` and `diff server` subcommand split.

dwelch-spike commented on cmd/diff.go:
```diff
@@ -32,95 +40,253 @@ var diffCmd = newDiffCmd()
+	return cmd
+}
+
+func newDiffServerCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "server [flags] <path/to/config>",
+		Short: "BETA: Diff a local config file against a running Aerospike server's configuration.",
+		Long: `BETA: Diff is used to compare a local configuration file against the configuration of a running Aerospike server. 
+				In this mode, only one config file path is required as an argument.
+				Note: The configuration file can be in yaml or conf format.`,
+		Example: `Diff a local .conf file against a running server
+  				asconfig diff server -h 127.0.0.1:3000 aerospike.conf`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			logger.Debug("Running server diff command")
+			return runServerDiff(cmd, args)
+		},
+	}
```
```suggestion
func newDiffServerCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "server [flags] <path/to/config>",
		Short: "BETA: Diff a local config file against a running Aerospike server's configuration.",
		Long: `BETA: Diff is used to compare a local configuration file against the configuration of a running Aerospike server. 
				This is useful for spotting drift between expected and actual Aerospike server configurations.
				In this mode, only one config file path is required as an argument.
				Note: The configuration file can be in yaml or conf format.`,
		Example: `Diff a local .conf file against a running server
  				asconfig diff server -h 127.0.0.1:3000 aerospike.conf`,
		RunE: func(cmd *cobra.Command, args []string) error {
			logger.Debug("Running server diff command")
			return runServerDiff(cmd, args)
		},
	}
```
A little note about how to use it might be good. Use this if you want.

a-spiker committed:
Update cmd/diff.go

update server diff usage example

Co-authored-by: dwelch-spike <53876192+dwelch-spike@users.noreply.github.com>

a-spiker committed:
Merge branch 'main' into TOOLS-3077

dwelch-spike reviewed (approved):
LGTM!
